torch/fx/experimental/symbolic_shapes.py

```diff
@@ -3665,6 +3674,43 @@ def _maybe_guard_eq(self, expr: Union["sympy.Eq", "sympy.Ne"], concrete_bool: bo
         free = sorted(free, key=lambda x: (self.size_hint(x, allow_none=True) or sys.maxsize, x.name), reverse=True)  # type: ignore[attr-defined]
         lhs = expr.lhs
         rhs = expr.rhs
+
+        if not is_eq:
+            # Unfortunately, range refinement is probably going to not
+            # work most of the time, because we don't support symbols
+            # in ranges.  For example, i0 <= s0 is un-rangeable, because
+            # we can't put s0 in the range.
+            #
+            # TODO: deal with conjunction/disjunction?
+            if len(free) == 1:
+                lhs = free[0]
+                r = try_solve(expr, lhs)
+                if r is not None and isinstance(r[0], (sympy.Lt, sympy.Le, sympy.Gt, sympy.Ge)) and isinstance(r[1], sympy.Integer):
+                    res = r[0]
+                    bound = int(r[1])
+                    if isinstance(res, sympy.Lt):
+                        # lhs < bound
+                        update_vr = [-sympy.oo, bound - 1]
+                    elif isinstance(res, sympy.Le):
+                        # lhs <= bound
+                        update_vr = [-sympy.oo, bound]
+                    elif isinstance(res, sympy.Gt):
+                        # lhs > bound
+                        update_vr = [bound + 1, sympy.oo]
+                    elif isinstance(res, sympy.Ge):
+                        # lhs >= bound
+                        update_vr = [bound, sympy.oo]
+                    else:
+                        raise AssertionError(f"illegal expr {expr}")
+                    old_range = self.var_to_range[lhs]
+                    self.var_to_range[lhs] &= ValueRanges(*update_vr)
+                    if old_range != self.var_to_range[lhs]:
+                        self.log.info("refined range on %s to %s", lhs, self.var_to_range[lhs])
+                        self._maybe_evaluate_static.cache_clear()
```

Comment:
It is basically the same, and a reasonable BE followup for me would be to eliminate one or the other. Honestly, the only reason there's a second copy is I didn't realize this was here. There are two major differences with the implementations:

* `refine_ranges` attempts to solve for every free variable, whereas `maybe_guard_rel` only propagates in the univariate case. This could potentially be a performance problem, especially when you have guards like `s0 + ... + s999 = s1000`, but it also increases precision. In the unbacked case, I actually do have a lot of guards like this, so I think the performance concern here is warranted.
* `refine_ranges` handles non-integral RHS (e.g., `x < f(y)`) by computing the bound of `f(y)` and then using the conservative end of the range. This is unconditionally good and worth using.